[blog] Automate Your Infrastructure with Automation API and C# (#5380)
pulumi/docs: content/blog/automation-api-dotnet/index.md

---
title: "Automate Your Infrastructure with Automation API and C#"
date: tbd
meta_desc: "tbd"
meta_image: automation_api.png
authors:

> - joshua-studt

Reviewer: FYI: you'll want to add a `.toml` file to `data/team/team` for this to render correctly.

---
title: "Automate Your Infrastructure with Automation API and C#"
date: tbd

> meta_desc: "tbd"

Reviewer 2: ```suggestion
meta_desc: "C# developers can programmatically build infrastructure (with out a CLI) using the Pulumi Automation API package. "
```

var result = await stack.UpAsync();
var bucketName = result.Outputs["bucket_name"];
```

You can also continue to use your existing Pulumi projects, and invoke them from the Automation API:

```csharp
var stackArgs = new LocalProgramArgs("stackName", "C:\path\to\pulumi\project\dir");
using var stack = await LocalWorkspace.CreateOrSelectStackAsync(stackArgs);

await stack.UpAsync();
```

And since the Automation API can be invoked and debugged just like any other code that you write, it enables you put together complex deployment workflows such as a blue-green deployment model.

```csharp
var workspace = await LocalWorkspace.CreateAsync(new LocalWorkspaceOptions
{
    Program = PulumiFn.Create<ApplicationStack>(), // use your existing Pulumi.Stack implementation
    ProjectSettings = new ProjectSettings("projectName", ProjectRuntimeName.Dotnet),
});

// your "blue" production code is already running

var green = await WorkspaceStack.SelectAsync("green", workspace);
await green.UpAsync();

// do your cutover to "green" logic
// and then teardown "blue"

var blue = await WorkspaceStack.SelectAsync("blue", workspace);
await blue.DestroyAsync();
```

> See the full Automation API examples in [C#](https://github.com/pulumi/automation-api-examples/tree/main/dotnet).

Reviewer 2: ```suggestion
See full Automation API examples in [C#](https://github.com/pulumi/automation-api-examples/tree/main/dotnet).
```

---
title: "Automate Your Infrastructure with Automation API and C#"

> date: tbd

Reviewer 2: ```suggestion
date: 2021-03-08
```

meta_desc: "C# developers can programmatically build infrastructure (with out a CLI) using the Pulumi Automation API package. "
meta_image: automation_api.png
authors:
- joshua-studt
tags:
- Automation API
- C#
- csharp
- dotnet
- .NET
---

{{% notes type="info" %}}
Joshua Studt is a Solutions Architect at Financial Independence Group and a Pulumi Community member who contributed the C# package for Automation API.

> {{% /notes %}

Reviewer 2: ```suggestion
{{% /notes %}}

```

Currently available in public preview, Pulumi's Automation API lets you to programmatically call upon the Pulumi engine to provision your infrastructure. Today, we are excited to announce C# support for Automation API, enabling .NET developers to automate infrastructure deployments, create complex orchestration workflows, build custom ops tooling, and build cloud frameworks. Read more about the Automation API [here]({{< relref "/blog/automation-api" >}}).

## Using Automation API in .NET

The `Pulumi.Automation` [NuGet package](https://www.nuget.org/packages/Pulumi.Automation) exposes a `LocalWorkspace` for  creating and managing Pulumi [Stacks]({{< relref "/docs/intro/concepts/stack" >}}), and a `WorkspaceStack` that is a programmatic representation of a Stack for updating, refreshing, previewing, and destroying cloud resources. The Automation API makes it trivial to run Pulumi programs inline:

> 

Reviewer 2: ```suggestion
```

await stack.Workspace.InstallPluginAsync("aws", "v3.30.1");
var result = await stack.UpAsync();
var bucketName = result.Outputs["bucket_name"];
```

You can use your existing Pulumi projects and invoke them from the Automation API:

```csharp
var stackArgs = new LocalProgramArgs("stackName", "C:\path\to\pulumi\project\dir");
using var stack = await LocalWorkspace.CreateOrSelectStackAsync(stackArgs);

await stack.UpAsync();
```

Since the Automation API can be invoked and debugged like any other code, it enables you to put together complex deployment workflows such as a blue-green deployment model.

```csharp
var workspace = await LocalWorkspace.CreateAsync(new LocalWorkspaceOptions
{
    Program = PulumiFn.Create<ApplicationStack>(), // use your existing Pulumi.Stack implementation
    ProjectSettings = new ProjectSettings("projectName", ProjectRuntimeName.Dotnet),
});

// your "blue" production code is already running

var green = await WorkspaceStack.SelectAsync("green", workspace);
await green.UpAsync();

// do your cutover to "green" logic
// and then teardown "blue"

var blue = await WorkspaceStack.SelectAsync("blue", workspace);
await blue.DestroyAsync();
```

> See full Automation API examples in [C#](https://github.com/pulumi/automation-api-examples/tree/main/dotnet).

Reviewer 2: ```suggestion
See full Automation API examples in [C#](https://github.com/pulumi/automation-api-examples/tree/main/dotnet). Let us know if you're using Automation API on [Twitter](https://twitter.com/PulumiCorp) or in [Pulumi Comunity Slack](https://slack.pulumi.com/).
```

---
title: "Automate Your Infrastructure with Automation API and C#"
date: 2021-03-08
meta_desc: "C# developers can programmatically build infrastructure (with out a CLI) using the Pulumi Automation API package. "
meta_image: automation_api.png
authors:
- joshua-studt
- sophia-parafina
tags:
- Automation API
- C#
- csharp
- dotnet
- .NET
---

{{% notes type="info" %}}
Joshua Studt is a Solutions Architect at Financial Independence Group and a Pulumi Community member who contributed the C# package for Automation API.
{{% /notes %}}

> Currently available in public preview, Pulumi's Automation API lets you to programmatically call upon the Pulumi engine to provision your infrastructure. Today, we are excited to announce C# support for Automation API, enabling .NET developers to automate infrastructure deployments, create complex orchestration workflows, build custom ops tooling, and build cloud frameworks. Read more about the Automation API [here]({{< relref "/blog/automation-api" >}}).

Reviewer 3: ```suggestion
Currently available in public preview, Pulumi's Automation API enables you to provision your infrastructure programmatically using the Pulumi engine. Today, we are excited to announce C# support for Automation API, enabling .NET developers to automate infrastructure deployments, create complex orchestration workflows, build custom ops tooling, and build cloud frameworks. Read more about the Automation API [here]({{< relref "/blog/automation-api" >}}).
```

await stack.Workspace.InstallPluginAsync("aws", "v3.30.1");
var result = await stack.UpAsync();
var bucketName = result.Outputs["bucket_name"];
```

You can use your existing Pulumi projects and invoke them from the Automation API:

```csharp
var stackArgs = new LocalProgramArgs("stackName", "C:\path\to\pulumi\project\dir");
using var stack = await LocalWorkspace.CreateOrSelectStackAsync(stackArgs);

await stack.UpAsync();
```

Since the Automation API can be invoked and debugged like any other code, it enables you to put together complex deployment workflows such as a blue-green deployment model:

```csharp
using var workspace = await LocalWorkspace.CreateAsync(new LocalWorkspaceOptions
{
    Program = PulumiFn.Create<ApplicationStack>(), // use your existing Pulumi.Stack implementation
    ProjectSettings = new ProjectSettings("projectName", ProjectRuntimeName.Dotnet),
});

// your "blue" production code is already running

var green = await WorkspaceStack.SelectAsync("green", workspace);
await green.UpAsync();

// do your cutover to "green" logic
// and then teardown "blue"

var blue = await WorkspaceStack.SelectAsync("blue", workspace);
await blue.DestroyAsync();
```

> See full Automation API examples in [C#](https://github.com/pulumi/automation-api-examples/tree/main/dotnet). Let us know if you're using Automation API on [Twitter](https://twitter.com/PulumiCorp) or in [Pulumi Comunity Slack](https://slack.pulumi.com/).

Reviewer 3: ```suggestion
See full Automation API examples in [C#](https://github.com/pulumi/automation-api-examples/tree/main/dotnet). Let us know if you're using Automation API on [Twitter](https://twitter.com/PulumiCorp) or in our [Community Slack](https://slack.pulumi.com/).
```

- .NET
---

{{% notes type="info" %}}
Joshua Studt is a Solutions Architect at Financial Independence Group and a Pulumi Community member who contributed the C# package for Automation API.
{{% /notes %}
Currently available in public preview, Pulumi's Automation API lets you to programmatically call upon the Pulumi engine to provision your infrastructure. Today, we are excited to announce C# support for Automation API, enabling .NET developers to automate infrastructure deployments, create complex orchestration workflows, build custom ops tooling, and build cloud frameworks. Read more about the Automation API [here.]({{< relref "/blog/automation-api" >}})

## Using Automation API in .NET

> The `Pulumi.Automation` [NuGet package](https://github.com/pulumi/pulumi/tree/master/sdk/dotnet/Pulumi.Automation) exposes a `LocalWorkspace` for  creating and managing Pulumi [Stacks]({{< relref "/docs/intro/concepts/stack" >}}), and a `WorkspaceStack` that is a programmatic representation of a Stack for updating, refreshing, previewing, and destroying cloud resources. The Automation API makes it trivial to run Pulumi programs inline:

Author: Yea that sounds better - done
